cmd/pulumi-test-language/providers/config_grpc_provider.go

```diff
@@ -0,0 +1,331 @@
+) (plugin.DiffConfigResponse, error) {
+	return plugin.DiffResult{}, nil
+}
+
+func (p *ConfigGrpcProvider) Invoke(
+	ctx context.Context, req plugin.InvokeRequest,
+) (plugin.InvokeResponse, error) {
+	switch {
+	case string(req.Tok) == fmt.Sprintf("%s:index:toSecret", p.Pkg()):
+		secreted := req.Args.Copy()
+		for k, v := range secreted {
+			secreted[k] = resource.MakeSecret(v)
+		}
+		return plugin.InvokeResponse{Properties: secreted}, nil
+	default:
+		return plugin.InvokeResponse{}, errors.New("Unknown function")
+	}
+}
+
+// This lower level implementation should be used at runtime specifically to test at the lower level, since not all
+// actual providers use plugin.Provider consistently but some are using older or modified versions.
+// pulumirpc.UnimplementedResourceProviderServer
+type configGrpcProviderServer struct {
+	pulumirpc.ResourceProviderServer
+
+	// Guard the state.
+	sync.Mutex
+
+	// State to capture of configuration-related requests.
+	configRequests []json.RawMessage
+}
+
+func (p *configGrpcProviderServer) logMessage(msg proto.Message) error {
+	req, err := protojson.Marshal(msg)
+	if err != nil {
+		return err
+	}
+	p.Mutex.Lock()
+	defer p.Mutex.Unlock()
+	type tagged struct {
+		Method  string          `json:"method"`
+		Message json.RawMessage `json:"message"`
+	}
+	bytes, err := json.Marshal(tagged{
+		Method:  string(msg.ProtoReflect().Descriptor().FullName()),
+		Message: req,
+	})
+	if err != nil {
+		return err
+	}
+	p.configRequests = append(p.configRequests, bytes)
+	return nil
+}
+
+func (p *configGrpcProviderServer) formatLoggedMessages() string {
+	p.Mutex.Lock()
+	defer p.Mutex.Unlock()
+	bytes, err := json.Marshal(p.configRequests)
+	contract.AssertNoErrorf(err, "json.Marshal failed")
+	return string(bytes)
+}
+
+func (p *configGrpcProviderServer) CheckConfig(
+	ctx context.Context,
+	req *pulumirpc.CheckRequest,
+) (*pulumirpc.CheckResponse, error) {
+	if err := p.logMessage(req); err != nil {
+		return nil, err
+	}
+	return p.ResourceProviderServer.CheckConfig(ctx, req)
+}
+
+func (p *configGrpcProviderServer) Configure(
+	ctx context.Context,
+	req *pulumirpc.ConfigureRequest,
+) (resp *pulumirpc.ConfigureResponse, err error) {
+	if err := p.logMessage(req); err != nil {
+		return nil, err
+	}
+	return p.ResourceProviderServer.Configure(ctx, req)
+}
+
+func (p *configGrpcProviderServer) DiffConfig(
+	ctx context.Context,
+	req *pulumirpc.DiffRequest,
+) (*pulumirpc.DiffResponse, error) {
+	if err := p.logMessage(req); err != nil {
+		return nil, err
+	}
+	return p.ResourceProviderServer.DiffConfig(ctx, req)
+}
+
+func (p *configGrpcProviderServer) Create(
+	ctx context.Context,
+	req *pulumirpc.CreateRequest,
+) (*pulumirpc.CreateResponse, error) {
+	id := ""
+	if !req.Preview {
+		id = "id0"
+	}
+	return &pulumirpc.CreateResponse{
+		Id: id,
+		Properties: &structpb.Struct{
+			Fields: map[string]*structpb.Value{
+				"config": structpb.NewStringValue(p.formatLoggedMessages()),
```

Comment:
Do we want to stringify this or is it possible to return the object as a "real" object? Or is this is a stupid question that totally defeats the point of the test because it'd stop us observing the encoding? :thinking:

Reply:
To return it as a real object we'd need another pass into gRPC structs,  worry of getting that "wrong" so I prefer just a sealed string envelope. I poked around a bit some more and cleaned up, I think it's ok-ish now; there is something unfortunate going on where the test framework does not partition the provider address space the way real Pulumi does. This started confounding these tests somehow. I need to index into which Configure/CheckConfig request to retrieve now.